### PR TITLE
MATX-303 : Material Node discovery 

### PR DESCRIPTION
**Pull request by kwokcb**
### Discover Logic.
- discovery returns any node or nodegraph which has a one or more `material` outputs
- Works on a nodegraph, materialassign, or document as the element to test from.
- Document level . It ignores implementation nodegraphs.
- Nodegraph discovery does not check if the graph is an implementation or not.

### Configurations Supported.
Refer to images in:
https://github.com/materialx/MaterialX/issues/737
for graphical view of configurations tested.

### Test Results.
*Material nodes discovered which are referenced by `materialassigns`*
- Found material node: top_level_material_assigned
- Found material node: top_level_material_in_graph_assigned/surfacematerial1
- Found material node: top_level_material_def_assigned

*Material nodes discover which are within a nodegraph*
- Found output material node: top_level_material_in_graph_no_assign/surfacematerial1
- Found output material node on definition: NG_material_def_material_1_0/surfacematerial1
- Found output material node: top_level_material_in_graph_assigned/surfacematerial1

*Material nodes or nodegraphs discovered within a Document*
This scans at the document level first, and then for each nodegraph scans inside. The difference
when scanning at document level is that nodegraph implementations are skipped so `NG_material_def_material_1_0/surfacematerial1` is not return.
- NodeGraph Node: top_level_material_in_graph_no_assign/surfacematerial1
- NodeGraph Node: top_level_material_in_graph_assigned/surfacematerial1
- Node: top_level_material_no_asssign
- Node: top_level_material_assigned
- Node: top_level_material_def
- Node: top_level_material_def_assigned

### CODE REVIEW

File: source/MaterialXCore/Material.cpp [hunk 1]
```diff
@@ -135,4 +139,118 @@ vector<OutputPtr> getConnectedOutputs(NodePtr node)
...
+                for (auto graphOutput : docNodeGraph->getOutputs())
+                {
+                    if (graphOutput->getType() == MATERIAL_TYPE_STRING)
+                    {
+                        NodePtr node = graphOutput->getConnectedNode();
+                        if (node && node->getType() == MATERIAL_TYPE_STRING &&
+                            !processedSources.count(docNodeGraph))
+                        {
+                            materialNodes.push_back(docNodeGraph);
+                            processedSources.insert(node);
```
[the remark this answers is not in the document]
Comment (kwokcb) [reply source/MaterialXCore/Material.cpp]: Thanks for the catch.

File: source/MaterialXCore/Material.cpp [hunk 2]
```diff
@@ -135,4 +139,118 @@ vector<OutputPtr> getConnectedOutputs(NodePtr node)
...
+                for (auto graphOutput : materialGraph->getOutputs())
+                {
+                    if (graphOutput->getType() == MATERIAL_TYPE_STRING)
+                    {
+                        NodePtr node = graphOutput->getConnectedNode();
+                        if (node && (node->getType() == MATERIAL_TYPE_STRING) && 
+                            !processedSources.count(node))
+                        {
+                            materialNodes.push_back(node);
+                            processedSources.insert(node);
```
[the remark this answers is not in the document]
Comment (kwokcb) [reply source/MaterialXCore/Material.cpp]: Yes. In agreement on this.

File: source/MaterialXCore/Material.h [hunk 3]
```diff
@@ -12,6 +12,7 @@
 #include <MaterialXCore/Export.h>
 
 #include <MaterialXCore/Node.h>
+#include <set>
```
Comment (niklasharrysson) [inline source/MaterialXCore/Material.h:15]: Is this include unused now?

**Review by niklasharrysson — APPROVED** [review]
Looks good, and I only had a small remark above.

**Pull request merged**